File: sys/src/slam/projection_factor.cc
#include "projection_factor.h"
#include "../base/rust.hpp"

namespace gtsam
{
    std::unique_ptr<SmartProjectionPoseFactor<gtsam::Cal3_S2>> new_smart_projection_pose_factor(
        const std::shared_ptr<noiseModel::Isotropic> & measurement_noise,
        const std::shared_ptr<Cal3_S2> & K,
        const Pose3 & sensor_P_body)
    {
        return std::make_unique<SmartProjectionPoseFactor<gtsam::Cal3_S2>>(
            to_boost_ptr(measurement_noise),
            to_boost_ptr(K),
            sensor_P_body);
    }

    void add(SmartProjectionPoseFactorCal3_S2 &factor, const Point2 & point, Key key) {
        factor.add(point, key);
    }
} // namespace gtsam

File: src/slam/projection_factor.rs
use cxx::{SharedPtr, UniquePtr};

use crate::{geometry::{cal3_s2::Cal3S2, point2::Point2, pose3::Pose3}, inference::key::IntoKey, linear::noise_model::IsotropicNoiseModel};

pub struct SmartProjectionPoseFactorCal3S2 {
    pub(crate) inner: UniquePtr<::sys::SmartProjectionPoseFactorCal3_S2>,
}

impl SmartProjectionPoseFactorCal3S2 {
    pub fn new(
        measurement_noise: &IsotropicNoiseModel,
        k: &Cal3S2,
        sensor_p_body: &Pose3,
    ) -> Self {
        Self {
            inner: ::sys::new_smart_projection_pose_factor(
                &measurement_noise.inner,
                &k.inner,
                &sensor_p_body.inner,
            ),
        }
    }

    pub fn add(
        self: &mut Self,
        point: &Point2,
        key: impl IntoKey
    ) {
        ::sys::add(
            self.inner.pin_mut(),
            &point.inner,
            key.into_key(),
        )
    }
}